sequencer/src/api/sql.rs

```diff
@@ -17,6 +31,80 @@ impl SequencerDataSource for DataSource {
 
         Ok(cfg.connect(provider).await?)
     }
+
+    async fn store_state<S: MerklizedState<SeqTypes>>(
+        &mut self,
+        path: MerklePath<S::Entry, S::Key, S::T>,
+        traversal_path: Vec<usize>,
+        block_number: u64,
+    ) -> anyhow::Result<()> {
+        <DataSource as UpdateStateData<SeqTypes, S>>::insert_merkle_nodes(
+            self,
+            path,
+            traversal_path,
+            block_number,
+        )
+        .await
+        .context("failed to insert merkle nodes! ")
+    }
+}
+
+#[async_trait]
+impl<D: SequencerDataSource + Send + Sync> UpdateStateStorage<SeqTypes, D> for ValidatedState {
+    async fn update_storage(
+        &self,
+        storage: &mut D,
+        leaf: &Leaf<SeqTypes>,
+        delta: Arc<Delta>,
+    ) -> anyhow::Result<()> {
+        let block_number = leaf.get_height();
+        let ValidatedState {
+            fee_merkle_tree,
+            block_merkle_tree,
+        } = self;
+
+        let Delta {
+            blocks_delta,
+            fees_delta,
+        } = delta.as_ref();
+
+        // Insert block merkle tree nodes
+        for delta in blocks_delta {
+            let (_, proof) = block_merkle_tree
+                .lookup(delta)
+                .expect_ok()
+                .context("Index not found in block merkle tree")?;
```

Comment:
I don't think deltas should ever be missing, but still I don't think we should panic if they are. Better to return an error, some state will be missing but we can recover from that once we implement state catchup.

Reply:
it never panics as the update function in query service just logs the errors

Reply:
Oh sorry, I got confused by `expect_ok()` (because `expect()` would be a panic. Disregrad